src/aspire/orientation/commonline_sync.py

```diff
@@ -1,34 +1,298 @@
 from aspire.orientation import Orient3D
 import logging
+import numpy as np
+
+import scipy.sparse as sps
+import scipy.sparse.linalg as spsl
 
 logger = logging.getLogger(__name__)
 
 
 class CommLineSync(Orient3D):
     """
-    Define a class to estimate 3D orientations using Synchronization described as below:
-    Y. Shkolnisky, and A. Singer, Viewing Direction Estimation in Cryo-EM Using Synchronization,
+    Define a class to estimate 3D orientations using synchronization matrix and voting method.
+
+    The related publications are listed as below:
+    Y. Shkolnisky, and A. Singer,
+    Viewing Direction Estimation in Cryo-EM Using Synchronization,
     SIAM J. Imaging Sciences, 5, 1088-1110 (2012).
 
+    A. Singer, R. R. Coifman, F. J. Sigworth, D. W. Chester, Y. Shkolnisky,
+    Detecting Consistent Common Lines in Cryo-EM by Voting,
+    Journal of Structural Biology, 169, 312-322 (2010).
     """
-    def __init__(self, src):
+    def __init__(self, src, nrad=None, ntheta=None):
+        """
+        Initialize an object for estimating 3D orientations using synchronization matrix
+
+        :param src: The source object of 2D denoised or class-averaged images with metadata
+        :param n_rad: The number of points in the radial direction
+        :param n_theta: The number of points in the theta direction
+        """
+        super().__init__(src, nrad=nrad, ntheta=ntheta)
+        self.syncmatrix = None
+
+    def estimate_rotations(self):
+        """
+        Estimate orientation matrices for all 2D images using synchronization matrix
+        """
+        if self.syncmatrix is None:
+            self.syncmatrix_vote()
+
+        s = self.syncmatrix
+        sz = s.shape
+        if len(sz) != 2:
+            raise ValueError('clmatrix must be a square matrix')
+        if sz[0] != sz[1]:
+            raise ValueError('clmatrix must be a square matrix')
+        if sz[0] % 2 == 1:
+            raise ValueError('clmatrix must be a square matrix of size 2Kx2K')
+
+        k = sz[0] // 2
```

Comment:
`k` is the number neighbouring images selected to calculate common lines. It can be smaller than the number of projections.  Renamed it to `n_check` to be consistent with the previous definition.